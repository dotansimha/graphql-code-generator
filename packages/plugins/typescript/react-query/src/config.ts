import { RawClientSideBasePluginConfig } from '@graphql-codegen/visitor-plugin-common';

export type HardcodedFetch = { endpoint: string; fetchParams?: Record<string, any> };
export type CustomFetch = { func: string; isReactHook?: boolean } | string;

/**
 * @description This plugin generates `React-Query` Hooks with TypeScript typings.
 *
 * It extends the basic TypeScript plugins: `@graphql-codegen/typescript`, `@graphql-codegen/typescript-operations` - and thus shares a similar configuration.
 */
export interface ReactQueryRawPluginConfig
  extends Omit<
    RawClientSideBasePluginConfig,
    | 'documentMode'
    | 'noGraphQLTag'
    | 'gqlImport'
    | 'documentNodeImport'
    | 'noExport'
    | 'importOperationTypesFrom'
    | 'importDocumentNodeExternallyFrom'
    | 'useTypeImports'
  > {
  /**
   * @description Customize the fetcher you wish to use in the generated file. React-Query is agnostic to the data-fetcing layer, so you should provide it, or use a custom one.
   *
   * The following options are available to use:
   * - 'fetch' - requires you to specify endpoint and headers on each call, and uses `fetch` to do the actual http call.
   * - `{ endpoint: string, fetchParams: RequestInit }`: hardcode your endpoint and fetch options into the generated output, using the environment `fetch` method. You can also use `process.env.MY_VAR` as endpoint or header value.
   * - `file#identifier` - You can use custom fetcher method that should implement the exported `ReactQueryFetcher` interface. Example: `./my-fetcher#myCustomFetcher`.
   * - `graphql-request`: Will generate each hook with `client` argument, where you should pass your own `GraphQLClient` (created from `graphql-request`).
   */
  fetcher?: 'fetch' | HardcodedFetch | 'graphql-request' | CustomFetch;

  /**
   * @default false
   * @description For each generate query hook adds `document` field with a
   * correspoding GraphQL query. Useful for `queryClient.fetchQuery`. Example:
   * queryClient.fetchQuery(
   *   useUserDetailsQuery.getKey(variables),
   *   () => gqlRequest(useUserDetailsQuery.document, variables),
   * )
   */
  exposeDocument?: boolean;

  /**
   * @default false
   * @description For each generate query hook adds getKey(variables: QueryVariables) function. Useful for cache updates. Example:
   * const query = useUserDetailsQuery(...);
   * const key = useUserDetailsQuery.getKey({id: theUsersId});
   * // use key in a cache update after a mutation
   */
  exposeQueryKeys?: boolean;

  /**
   * @default unknown
   * @description Changes the default "TError" generic type.
   */
  errorType?: string;
}
