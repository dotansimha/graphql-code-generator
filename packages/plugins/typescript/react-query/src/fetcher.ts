import { OperationDefinitionNode } from 'graphql';

export interface FetcherRenderer {
  generateFetcherImplementaion: () => string;
  generateQueryHook: (
    node: OperationDefinitionNode,
    documentVariableName: string,
    operationName: string,
    operationResultType: string,
    operationVariablesTypes: string,
    hasRequiredVariables: boolean
  ) => string;
  generateMutationHook: (
    node: OperationDefinitionNode,
    documentVariableName: string,
    operationName: string,
    operationResultType: string,
    operationVariablesTypes: string
  ) => string;
}
