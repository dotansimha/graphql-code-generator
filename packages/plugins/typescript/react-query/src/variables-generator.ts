import { OperationDefinitionNode } from 'graphql';

export function generateQueryVariablesSignature(
  hasRequiredVariables: boolean,
  operationVariablesTypes: string
): string {
  return `variables${hasRequiredVariables ? '' : '?'}: ${operationVariablesTypes}`;
}

export function generateQueryKey(node: OperationDefinitionNode): string {
  return `['${node.name.value}', variables]`;
}

export function generateQueryKeyMaker(
  node: OperationDefinitionNode,
  operationName: string,
  operationVariablesTypes: string,
  hasRequiredVariables: boolean
) {
  const signature = generateQueryVariablesSignature(hasRequiredVariables, operationVariablesTypes);
  return `\nuse${operationName}.getKey = (${signature}) => ${generateQueryKey(node)};\n`;
}
