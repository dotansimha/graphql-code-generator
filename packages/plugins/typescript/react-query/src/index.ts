import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { visit, GraphQLSchema, concatAST, Kind, FragmentDefinitionNode } from 'graphql';
import { LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { ReactQueryVisitor } from './visitor';
import { extname } from 'path';
import { ReactQueryRawPluginConfig } from './config';

export const plugin: PluginFunction<ReactQueryRawPluginConfig, Types.ComplexPluginOutput> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: ReactQueryRawPluginConfig
) => {
  const allAst = concatAST(documents.map(v => v.document));

  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];

  const visitor = new ReactQueryVisitor(schema, allFragments, config, documents);
  const visitorResult = visit(allAst, { leave: visitor });

  if (visitor.hasOperations) {
    return {
      prepend: [...visitor.getImports(), visitor.getFetcherImplementation()],
      content: [visitor.fragments, ...visitorResult.definitions.filter(t => typeof t === 'string')].join('\n'),
    };
  }

  return {
    prepend: [...visitor.getImports()],
    content: [visitor.fragments, ...visitorResult.definitions.filter(t => typeof t === 'string')].join('\n'),
  };
};

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: ReactQueryVisitor,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts' && extname(outputFile) !== '.tsx') {
    throw new Error(`Plugin "typescript-react-query" requires extension to be ".ts" or ".tsx"!`);
  }
};

export { ReactQueryVisitor };
