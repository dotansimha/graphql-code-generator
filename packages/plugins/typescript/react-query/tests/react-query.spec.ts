import { validateTs } from '@graphql-codegen/testing';
import { parse, buildClientSchema, GraphQLSchema } from 'graphql';
import { plugin } from '../src';
import { Types, mergeOutputs } from '@graphql-codegen/plugin-helpers';
import { plugin as tsPlugin } from '../../typescript/src/index';
import { plugin as tsDocumentsPlugin } from '../../operations/src/index';

const validateTypeScript = async (
  output: Types.PluginOutput,
  testSchema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: any
) => {
  const tsOutput = await tsPlugin(testSchema, documents, config, { outputFile: '' });
  const tsDocumentsOutput = await tsDocumentsPlugin(testSchema, documents, config, { outputFile: '' });
  const merged = mergeOutputs([tsOutput, tsDocumentsOutput, output]);
  validateTs(merged, undefined, true, false);

  return merged;
};

describe('React-Query', () => {
  const schema = buildClientSchema(require('../../../../../dev-test/githunt/schema.json'));
  const basicDoc = parse(/* GraphQL */ `
    query test {
      feed {
        id
        commentCount
        repository {
          full_name
          html_url
          owner {
            avatar_url
          }
        }
      }
    }
  `);
  const basicMutation = parse(/* GraphQL */ `
    mutation test($name: String) {
      submitRepository(repoFullName: $name) {
        id
      }
    }
  `);
  const basicFragment = parse(/* GraphQL */ `
    fragment EntryData on Entry {
      feed {
        id
        commentCount
        repository {
          full_name
          html_url
          owner {
            avatar_url
          }
        }
      }
    }
  `);
  const docs = [
    {
      document: basicDoc,
    },
    {
      document: basicMutation,
    },
  ];
  const notOperationDocs = [
    {
      document: basicFragment,
    },
  ];

  it('should allow to override TError type', async () => {
    const config = {
      errorType: 'any',
    };

    const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;
    expect(out.content).not.toContain(`TError = unknown`);
    expect(out.content).toContain(`TError = any`);
  });

  describe('fetcher: custom-mapper', () => {
    it('Should generate query correctly with external mapper', async () => {
      const config = {
        fetcher: './my-file#myCustomFetcher',
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend).toContain(
        `import { useQuery, UseQueryOptions, useMutation, UseMutationOptions } from 'react-query';`
      );
      expect(out.prepend).toContain(`import { myCustomFetcher } from './my-file';`);
      expect(out.content).toBeSimilarStringTo(`export const useTestQuery = <
          TData = TTestQuery,
          TError = unknown
        >(
          variables?: TTestQueryVariables, 
          options?: UseQueryOptions<TTestQuery, TError, TData>
        ) => 
        useQuery<TTestQuery, TError, TData>(
          ['test', variables],
          myCustomFetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
          options
        );`);
      expect(out.content).toBeSimilarStringTo(`export const useTestMutation = <
        TError = unknown,
        TContext = unknown
      >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
      useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
        (variables?: TTestMutationVariables) => myCustomFetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
        options
      );`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });

    it('Should generate query correctly with internal mapper', async () => {
      const config = {
        fetcher: 'myCustomFetcher',
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend).toContain(
        `import { useQuery, UseQueryOptions, useMutation, UseMutationOptions } from 'react-query';`
      );
      expect(out.content).toBeSimilarStringTo(`export const useTestQuery = <
        TData = TTestQuery,
        TError = unknown
      >(
        variables?: TTestQueryVariables, 
        options?: UseQueryOptions<TTestQuery, TError, TData>
      ) => 
      useQuery<TTestQuery, TError, TData>(
        ['test', variables],
        myCustomFetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
        options
      );`);

      expect(out.content).toBeSimilarStringTo(`export const useTestMutation = <
        TError = unknown,
        TContext = unknown
      >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
      useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
        (variables?: TTestMutationVariables) => myCustomFetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
        options
      );`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });

    it('Should generate mutation correctly with lazy variables', async () => {
      const config = {
        fetcher: {
          func: './my-file#useCustomFetcher',
          isReactHook: true,
        },
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend).toContain(
        `import { useQuery, UseQueryOptions, useMutation, UseMutationOptions } from 'react-query';`
      );
      expect(out.prepend).toContain(`import { useCustomFetcher } from './my-file';`);
      expect(out.content).toBeSimilarStringTo(`export const useTestQuery = <
          TData = TTestQuery,
          TError = unknown
        >(
          variables?: TTestQueryVariables, 
          options?: UseQueryOptions<TTestQuery, TError, TData>
        ) => 
        useQuery<TTestQuery, TError, TData>(
          ['test', variables],
          useCustomFetcher<TTestQuery, TTestQueryVariables>(TestDocument).bind(null, variables),
          options
        );`);
      expect(out.content).toBeSimilarStringTo(`export const useTestMutation = <
        TError = unknown,
        TContext = unknown
      >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
      useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
        useCustomFetcher<TTestMutation, TTestMutationVariables>(TestDocument),
        options
      );`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });
  });

  describe('fetcher: graphql-request', () => {
    it('Should generate query correctly with client', async () => {
      const config = {
        fetcher: 'graphql-request',
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend).toContain(
        `import { useQuery, UseQueryOptions, useMutation, UseMutationOptions } from 'react-query';`
      );
      expect(out.prepend).toContain(`import { GraphQLClient } from 'graphql-request';`);
      expect(out.prepend[2])
        .toBeSimilarStringTo(`    function fetcher<TData, TVariables>(client: GraphQLClient, query: string, variables?: TVariables) {
          return async (): Promise<TData> => client.request<TData, TVariables>(query, variables);
        }`);
      expect(out.content).toBeSimilarStringTo(`export const useTestQuery = <
          TData = TTestQuery,
          TError = unknown
        >(
          client: GraphQLClient, 
          variables?: TTestQueryVariables, 
          options?: UseQueryOptions<TTestQuery, TError, TData>
        ) => 
        useQuery<TTestQuery, TError, TData>(
          ['test', variables],
          fetcher<TTestQuery, TTestQueryVariables>(client, TestDocument, variables),
          options
        );`);
      expect(out.content).toBeSimilarStringTo(`export const useTestMutation = <
        TError = unknown,
        TContext = unknown
      >(
        client: GraphQLClient, 
        options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>
      ) => 
      useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
        (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(client, TestDocument, variables)(),
        options
      );`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });
    it('Should support useTypeImports', async () => {
      const config = {
        fetcher: 'graphql-request',
        useTypeImports: true,
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend).toContain(`import type { GraphQLClient } from 'graphql-request';`);
    });
  });

  describe('fetcher: hardcoded-fetch', () => {
    it('Should generate query correctly with hardcoded endpoint', async () => {
      const config = {
        fetcher: {
          endpoint: 'http://localhost:3000/graphql',
        },
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend).toContain(
        `import { useQuery, UseQueryOptions, useMutation, UseMutationOptions } from 'react-query';`
      );
      expect(out.prepend[1])
        .toBeSimilarStringTo(`    function fetcher<TData, TVariables>(query: string, variables?: TVariables) {
        return async (): Promise<TData> => {
          const res = await fetch("http://localhost:3000/graphql", {
            method: "POST",
            body: JSON.stringify({ query, variables }),
          });
      
          const json = await res.json();
      
          if (json.errors) {
            const { message } = json.errors[0];
      
            throw new Error(message);
          }
      
          return json.data;
        }
      }`);
      expect(out.content).toBeSimilarStringTo(`export const useTestQuery = <
        TData = TTestQuery,
        TError = unknown
      >(
        variables?: TTestQueryVariables, 
        options?: UseQueryOptions<TTestQuery, TError, TData>
      ) => 
      useQuery<TTestQuery, TError, TData>(
        ['test', variables],
        fetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
        options
      );`);

      expect(out.content).toBeSimilarStringTo(`export const useTestMutation = <
        TError = unknown,
        TContext = unknown
      >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
      useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
        (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
        options
      );`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });

    it('Should generate query correctly with fetch config', async () => {
      const config = {
        fetcher: {
          endpoint: 'http://localhost:3000/graphql',
          fetchParams: {
            headers: {
              Authorization: 'Bearer XYZ',
            },
          },
        },
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend[1])
        .toBeSimilarStringTo(`function fetcher<TData, TVariables>(query: string, variables?: TVariables) {
        return async (): Promise<TData> => {
          const res = await fetch("http://localhost:3000/graphql", {
            method: "POST",
            headers: {"Authorization":"Bearer XYZ"},
            body: JSON.stringify({ query, variables }),
          });
      
          const json = await res.json();
      
          if (json.errors) {
            const { message } = json.errors[0];
      
            throw new Error(message);
          }
      
          return json.data;
        }
      }`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });

    it('Should generate query correctly with hardcoded endpoint from env var', async () => {
      const config = {
        fetcher: {
          endpoint: 'process.env.ENDPOINT_URL',
        },
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend[1])
        .toBeSimilarStringTo(`function fetcher<TData, TVariables>(query: string, variables?: TVariables) {
        return async (): Promise<TData> => {
          const res = await fetch(process.env.ENDPOINT_URL as string, {
            method: "POST",
            body: JSON.stringify({ query, variables }),
          });
      
          const json = await res.json();
      
          if (json.errors) {
            const { message } = json.errors[0];
      
            throw new Error(message);
          }
      
          return json.data;
        }
      }`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });

    it('Should generate query correctly with hardcoded endpoint from just identifier', async () => {
      const config = {
        fetcher: {
          endpoint: 'myEndpoint',
        },
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend[1])
        .toBeSimilarStringTo(`    function fetcher<TData, TVariables>(query: string, variables?: TVariables) {
        return async (): Promise<TData> => {
          const res = await fetch(myEndpoint as string, {
            method: "POST",
            body: JSON.stringify({ query, variables }),
          });
      
          const json = await res.json();
      
          if (json.errors) {
            const { message } = json.errors[0];
      
            throw new Error(message);
          }
      
          return json.data;
        }
      }`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });
  });

  describe('fetcher: fetch', () => {
    it('Should generate query and mutation correctly', async () => {
      const config = {
        fetcher: 'fetch',
        typesPrefix: 'T',
      };

      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;

      expect(out.prepend).toContain(
        `import { useQuery, UseQueryOptions, useMutation, UseMutationOptions } from 'react-query';`
      );

      expect(out.content).toBeSimilarStringTo(`export const useTestQuery = <
        TData = TTestQuery,
        TError = unknown
      >(
        dataSource: { endpoint: string, fetchParams?: RequestInit }, 
        variables?: TTestQueryVariables, 
        options?: UseQueryOptions<TTestQuery, TError, TData>
      ) => 
      useQuery<TTestQuery, TError, TData>(
        ['test', variables],
        fetcher<TTestQuery, TTestQueryVariables>(dataSource.endpoint, dataSource.fetchParams || {}, TestDocument, variables),
        options
      );`);

      expect(out.content).toBeSimilarStringTo(`export const useTestMutation = <
          TError = unknown,
          TContext = unknown
        >(
          dataSource: { endpoint: string, fetchParams?: RequestInit }, 
          options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>
        ) => 
        useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
          (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(dataSource.endpoint, dataSource.fetchParams || {}, TestDocument, variables)(),
          options
        );`);

      expect(out.content).toMatchSnapshot();
      await validateTypeScript(mergeOutputs(out), schema, docs, config);
    });
  });

  describe('exposeDocument: true', () => {
    it('Should generate document field for each query', async () => {
      const config = {
        fetcher: 'fetch',
        exposeDocument: true,
      };
      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;
      expect(out.content).toBeSimilarStringTo(`useTestQuery.document = TestDocument;`);
    });
  });

  describe('exposeQueryKeys: true', () => {
    it('Should generate getKey for each query', async () => {
      const config = {
        fetcher: 'fetch',
        exposeQueryKeys: true,
      };
      const out = (await plugin(schema, docs, config)) as Types.ComplexPluginOutput;
      expect(out.content).toBeSimilarStringTo(
        `useTestQuery.getKey = (variables?: TestQueryVariables) => ['test', variables];`
      );
    });
  });

  it('Should not generate fetcher if there are no operations', async () => {
    const out = (await plugin(schema, notOperationDocs, {})) as Types.ComplexPluginOutput;
    expect(out.prepend).not.toBeSimilarStringTo(`function fetcher<TData, TVariables>(`);

    const config = {
      fetcher: 'graphql-request',
    };

    const outGraphqlRequest = (await plugin(schema, notOperationDocs, config)) as Types.ComplexPluginOutput;
    expect(outGraphqlRequest.prepend).not.toContain(`import { GraphQLClient } from 'graphql-request';`);
  });
});
