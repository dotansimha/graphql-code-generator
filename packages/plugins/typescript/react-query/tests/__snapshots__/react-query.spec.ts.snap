// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React-Query fetcher: custom-mapper Should generate mutation correctly with lazy variables 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      useCustomFetcher<TTestQuery, TTestQueryVariables>(TestDocument).bind(null, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      useCustomFetcher<TTestMutation, TTestMutationVariables>(TestDocument),
      options
    );"
`;

exports[`React-Query fetcher: custom-mapper Should generate query correctly with external mapper 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      myCustomFetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      (variables?: TTestMutationVariables) => myCustomFetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
      options
    );"
`;

exports[`React-Query fetcher: custom-mapper Should generate query correctly with internal mapper 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      myCustomFetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      (variables?: TTestMutationVariables) => myCustomFetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
      options
    );"
`;

exports[`React-Query fetcher: fetch Should generate query and mutation correctly 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      dataSource: { endpoint: string, fetchParams?: RequestInit }, 
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      fetcher<TTestQuery, TTestQueryVariables>(dataSource.endpoint, dataSource.fetchParams || {}, TestDocument, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(
      dataSource: { endpoint: string, fetchParams?: RequestInit }, 
      options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>
    ) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(dataSource.endpoint, dataSource.fetchParams || {}, TestDocument, variables)(),
      options
    );"
`;

exports[`React-Query fetcher: graphql-request Should generate query correctly with client 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      client: GraphQLClient, 
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      fetcher<TTestQuery, TTestQueryVariables>(client, TestDocument, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(
      client: GraphQLClient, 
      options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>
    ) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(client, TestDocument, variables)(),
      options
    );"
`;

exports[`React-Query fetcher: hardcoded-fetch Should generate query correctly with fetch config 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      fetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
      options
    );"
`;

exports[`React-Query fetcher: hardcoded-fetch Should generate query correctly with hardcoded endpoint 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      fetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
      options
    );"
`;

exports[`React-Query fetcher: hardcoded-fetch Should generate query correctly with hardcoded endpoint from env var 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      fetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
      options
    );"
`;

exports[`React-Query fetcher: hardcoded-fetch Should generate query correctly with hardcoded endpoint from just identifier 1`] = `
"
export const TestDocument = \`
    query test {
  feed {
    id
    commentCount
    repository {
      full_name
      html_url
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const useTestQuery = <
      TData = TTestQuery,
      TError = unknown
    >(
      variables?: TTestQueryVariables, 
      options?: UseQueryOptions<TTestQuery, TError, TData>
    ) => 
    useQuery<TTestQuery, TError, TData>(
      ['test', variables],
      fetcher<TTestQuery, TTestQueryVariables>(TestDocument, variables),
      options
    );
export const TestDocument = \`
    mutation test($name: String) {
  submitRepository(repoFullName: $name) {
    id
  }
}
    \`;
export const useTestMutation = <
      TError = unknown,
      TContext = unknown
    >(options?: UseMutationOptions<TTestMutation, TError, TTestMutationVariables, TContext>) => 
    useMutation<TTestMutation, TError, TTestMutationVariables, TContext>(
      (variables?: TTestMutationVariables) => fetcher<TTestMutation, TTestMutationVariables>(TestDocument, variables)(),
      options
    );"
`;
